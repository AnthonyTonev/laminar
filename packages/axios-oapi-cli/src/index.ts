export { convertSchema } from './convert-schema';
export { oapiTs, schemaTs } from './convert';
