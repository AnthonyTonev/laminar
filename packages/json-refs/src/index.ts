import fetch from 'node-fetch';
import { URL } from 'url';

interface IdSchema {
  $id: string;
}
interface RefSchema {
  $ref: string;
}
interface TraversableSchema {
  [key: string]: any;
}

interface FilesMap {
  [file: string]: any;
}

interface Context {
  schema: any;
  files: FilesMap;
}

export const isTraversable = (schema: any): schema is TraversableSchema =>
  schema && typeof schema === 'object';

export const isIdSchema = (schema: any): schema is IdSchema =>
  isTraversable(schema) && '$id' in schema && schema.$id;

export const isRefSchema = (schema: any): schema is RefSchema =>
  isTraversable(schema) && '$ref' in schema && schema.$ref;

export const currentId = (schema: any, parentId?: string) =>
  isIdSchema(schema) ? new URL(schema.$id, parentId).toString() : parentId;

export const currentUrl = (url: string | undefined, id?: string) =>
  url ? new URL(url, id).toString() : undefined;

export const reduceSchema = <TResult = any>(
  schema: any,
  cb: (all: TResult, item: any, id?: string) => TResult,
  initial: TResult,
  id?: string,
): TResult =>
  isTraversable(schema)
    ? Object.values(schema).reduce(
        (all, item) => reduceSchema(item, cb, all, currentId(schema, id)),
        cb(initial, schema, id),
      )
    : cb(initial, schema, id);

export const extractNamedRefs = (document: any): FilesMap =>
  reduceSchema(
    document,
    (all, item, id) =>
      isIdSchema(item) ? { ...all, [new URL(item.$id, id).toString()]: item } : all,
    {},
  );

export const extractUrls = (schema: any, namedRefs: string[] = []) =>
  reduceSchema<string[]>(
    schema,
    (all, item, id) => {
      if (isRefSchema(item)) {
        const [url] = item.$ref.split('#');
        const fullUrl = currentUrl(url, id);

        if (fullUrl && !all.includes(fullUrl) && !namedRefs.includes(fullUrl)) {
          return [...all, fullUrl];
        }
      }
      return all;
    },
    [],
  );

export const extractFiles = async (schema: any): Promise<FilesMap> => {
  const refs = extractNamedRefs(schema);
  const result = await Promise.all(
    extractUrls(schema, Object.keys(refs)).map(url =>
      fetch(url)
        .then(response => response.json())
        .then(content => extractFiles(content).then(filesMap => ({ ...filesMap, [url]: content }))),
    ),
  );
  return result.reduce((all, item) => ({ ...all, ...item }), refs);
};

export const parseJsonPointer = (name: string) =>
  decodeURIComponent(name.replace('~1', '/').replace('~0', '~'));

export const getJsonPointer = (document: any, pointer: string) =>
  pointer
    .split('/')
    .reduce<any>(
      (item, name) => (name ? (item ? item[parseJsonPointer(name)] : undefined) : item),
      document,
    );

export const resolveNestedRefs = (schema: any, context: Context, parentId?: string) => {
  const id = currentId(schema, parentId);

  if (isTraversable(schema)) {
    for (const [key, item] of Object.entries(schema)) {
      schema[key] = resolveNestedRefs(item, context, id);
    }
  }

  if (isRefSchema(schema)) {
    const [url, pointer] = schema.$ref.split('#');
    const fullUrl = currentUrl(url, id);
    const currentDocument = fullUrl ? context.files[fullUrl] : context.schema;

    return pointer ? getJsonPointer(currentDocument, pointer) : currentDocument;
  }

  return schema;
};

export const resolveRefs = async (schema: any): Promise<any> =>
  resolveNestedRefs(schema, { schema, files: await extractFiles(schema) });
