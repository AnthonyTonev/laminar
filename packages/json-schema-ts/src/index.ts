const parse = (schema: any, refs: { [key: string]: string }) => {
  if (schema.type === 'string' || 'pattern' in schema) {
    return 'string';
  }

  if (schema.type === 'integer' || schema.type === 'number') {
    return 'number';
  }
};
