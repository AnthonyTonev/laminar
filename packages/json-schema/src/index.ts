import { resolveRefs } from '@ovotech/json-refs';
import { validateSchema } from './helpers';
import { messages } from './messages';
import { Schema, ValidateOptions } from './types';
import { validateAdditionalProperties } from './validators/validateAdditionalProperties';
import { validateAllOf } from './validators/validateAllOf';
import { validateAnyOf } from './validators/validateAnyOf';
import { validateBoolean } from './validators/validateBoolean';
import { validateConditionals } from './validators/validateConditionals';
import { validateConst } from './validators/validateConst';
import { validateContains } from './validators/validateContains';
import { validateDependencies } from './validators/validateDependencies';
import { validateEnum } from './validators/validateEnum';
import { validateExclusiveMaximum } from './validators/validateExclusiveMaximum';
import { validateExclusiveMinimum } from './validators/validateExclusiveMinimum';
import { validateFormat } from './validators/validateFormat';
import { validateItems } from './validators/validateItems';
import { validateMaximum } from './validators/validateMaximum';
import { validateMaxItems } from './validators/validateMaxItems';
import { validateMaxLength } from './validators/validateMaxLength';
import { validateMaxProperties } from './validators/validateMaxProperties';
import { validateMinimum } from './validators/validateMinimum';
import { validateMinItems } from './validators/validateMinItems';
import { validateMinLength } from './validators/validateMinLength';
import { validateMinProperties } from './validators/validateMinProperties';
import { validateMultipleOf } from './validators/validateMultipleOf';
import { validateNot } from './validators/validateNot';
import { validateOneOf } from './validators/validateOneOf';
import { validatePattern } from './validators/validatePattern';
import { validatePatternProperties } from './validators/validatePatternProperties';
import { validateProperties } from './validators/validateProperties';
import { validatePropertyNames } from './validators/validatePropertyNames';
import { validateRequired } from './validators/validateRequired';
import { validateType } from './validators/validateType';
import { validateUniqueItems } from './validators/validateUniqueItems';

export {
  childOptions,
  isJsonSchema,
  isObject,
  isEqual,
  isUniqueWith,
  NoErrors,
  HasError,
  HasErrors,
  CombineResults,
  validateSchema,
} from './helpers';

export { Schema, JsonSchema, ValidateOptions, Validator, Messages } from './types';

export const draft7 = [
  validateBoolean,
  validateConst,
  validateEnum,
  validateType,
  validateFormat,
  validateMultipleOf,
  validateMinimum,
  validateExclusiveMinimum,
  validateMaximum,
  validateExclusiveMaximum,
  validatePattern,
  validateMinLength,
  validateMaxLength,
  validateMinProperties,
  validateMaxProperties,
  validateRequired,
  validateProperties,
  validatePatternProperties,
  validatePropertyNames,
  validateDependencies,
  validateAdditionalProperties,
  validateMinItems,
  validateMaxItems,
  validateUniqueItems,
  validateItems,
  validateContains,
  validateConditionals,
  validateNot,
  validateOneOf,
  validateAllOf,
  validateAnyOf,
];

export const validate = async (
  original: Schema,
  value: any,
  { name = 'value', validators = draft7 }: Partial<ValidateOptions> = {},
) => {
  const schema = await resolveRefs(original);
  const result = validateSchema(schema, value, { name, validators });
  const errors = result.errors.map(error => messages[error.code](error));

  return { schema, errors, valid: errors.length === 0 };
};
