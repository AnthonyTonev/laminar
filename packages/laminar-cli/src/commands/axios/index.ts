import { Schema } from '@ovotech/json-schema';
import * as commander from 'commander';
import * as fs from 'fs';
import * as YAML from 'yaml';
import { printDocument } from '@ovotech/ts-compose';
import { convertOapi } from './convert';
import { green, yellow } from 'chalk';
import { dirname } from 'path';
import { Logger } from '../../types';
import { toContext, toString } from '../../helpers';

export const processFile = async (file: string): Promise<string> => {
  const { context, value } = await toContext(file);
  return printDocument(convertOapi(context, value));
};

const parseSchema = (type: string, content: string): Schema => {
  switch (type) {
    case 'json':
      return JSON.parse(content);
    case 'yaml':
      return YAML.parse(content);
    default:
      throw new Error(`Unknown STDIN type: ${type}, accepts only "json" and "yaml"`);
  }
};

export const axiosCommand = (logger: Logger = console): commander.Command =>
  commander
    .createCommand('axios')
    .description('Convert openapi schemas to typescript types for axios')
    .option('-f, --file <file>', 'Filename to process, uses STDIN if not specified')
    .option('-t, --stdin-type <type>', 'Type of STDIN data: "json" or "yaml"', 'json')
    .option('-o, --output <output>', 'File to output to, uses STDOUT if not specified')
    .action(async ({ file, output, stdinType }) => {
      const input = file ? file : parseSchema(stdinType, await toString(process.stdin));
      const result = await processFile(input);

      if (output) {
        fs.mkdirSync(dirname(output), { recursive: true });
        logger.info(
          `OpanAPI Schema ${green(file ? file : 'STDIN')} -> ${yellow(output)} axios types`,
        );
        fs.writeFileSync(output, result);
      } else {
        process.stdout.write(result);
      }
    });
