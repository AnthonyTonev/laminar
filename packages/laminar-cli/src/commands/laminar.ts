import * as commander from 'commander';
import { apiCommand } from './api';
import { axiosCommand } from './axios';
import { jsonSchemaCommand } from './json-shema';
import { Logger } from '../types';

export const laminarCommand = (output: Logger = console): commander.Command =>
  commander
    .createCommand('laminar')
    .version('0.9.0')
    .description(
      `Laminar CLI - a tool for converting OpenApi Schemas to appropriate typescript types

laminar api
  Convert openapi schema to types used by laminar api (openApiTyped middleware)

lamianr axios
  Convert openapi schema to types used by axios

lamianr json-schema
  Convert json schema to types

Example:
  laminar api --file openapi.yaml --output src/__generated__/openapi.yaml.ts
  curl http://example.com/openapi.yaml | laminar api --stdin-type yaml | prettier --stdin-filepath openapi.yamlts > src/__generated__/openapi.yaml.ts
  laminar api --file openapi.yaml --output src/__generated__/openapi.yaml.ts --watch
`,
    )
    .addCommand(apiCommand(output))
    .addCommand(axiosCommand(output))
    .addCommand(jsonSchemaCommand(output));
