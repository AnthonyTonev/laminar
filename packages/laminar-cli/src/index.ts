import { ResolvedSchema } from '@ovotech/json-schema';
import { printDocument } from '@ovotech/ts-compose';
import { OpenAPIObject, SchemaObject } from 'openapi3-ts';
import { convertOapi as apiConvert } from './commands/api/convert';
import { convertOapi as axiosConvert } from './commands/axios/convert';
import { convertSchema } from './convert-schema';

export { convertSchema } from './convert-schema';

export const apiContent = ({ schema, refs }: ResolvedSchema): string =>
  printDocument(apiConvert({ root: schema as OpenAPIObject, refs }, schema as OpenAPIObject));

export const axiosContent = ({ schema, refs }: ResolvedSchema): string =>
  printDocument(axiosConvert({ root: schema as OpenAPIObject, refs }, schema as OpenAPIObject));

export const schemaContent = ({ schema, refs }: ResolvedSchema): string =>
  printDocument(convertSchema({ root: schema as SchemaObject, refs }, schema));
