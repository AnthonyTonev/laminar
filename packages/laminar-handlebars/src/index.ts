import { Middleware, Response, response } from '@ovotech/laminar';
import { readdirSync, readFileSync, existsSync } from 'fs';
import { compile, RuntimeOptions, TemplateDelegate } from 'handlebars';
import { join } from 'path';
import { OutgoingHttpHeaders } from 'http';

export interface TemplateConfig {
  helpers?: RuntimeOptions['helpers'];
  dir?: string;
  partials?: string;
  extension?: string;
  views?: string;
  headers?: OutgoingHttpHeaders;
}

export interface CompileTemplatesOptions {
  childDir: string;
  dir: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  compileOptions?: any;
  extension: string;
}

export interface Templates {
  [key: string]: TemplateDelegate;
}

export type HandlebarsRender = (
  view: string,
  data?: Record<string, unknown>,
  options?: Partial<Response>,
) => Response;

export interface RequestHandlebars {
  render: HandlebarsRender;
}

export const deepReaddirSync = (dir: string, childDir: string, parent = ''): string[] =>
  readdirSync(join(dir, childDir), { withFileTypes: true }).reduce<string[]>(
    (all, item) =>
      item.isDirectory()
        ? [...all, ...deepReaddirSync(dir, join(childDir, item.name), join(parent, item.name))]
        : [...all, join(parent, item.name)],
    [],
  );

export const compileTemplates = ({
  childDir,
  dir,
  extension,
  compileOptions,
}: CompileTemplatesOptions): Templates =>
  deepReaddirSync(dir, childDir)
    .filter((file) => file.endsWith(extension))
    .reduce((all, file) => {
      const name = file.slice(0, -(extension.length + 1));
      const input = readFileSync(join(dir, childDir, file), 'utf8');
      return { ...all, [name]: compile(input, compileOptions) };
    }, {});

export const handlebarsMiddleware = ({
  helpers = {},
  dir = process.cwd(),
  extension = 'handlebars',
  partials = 'partials',
  views = 'views',
  headers,
}: TemplateConfig): Middleware<RequestHandlebars> => {
  const defaultHeaders = { 'content-type': 'text/html', ...headers };

  const partialTemplates = existsSync(join(dir, partials))
    ? compileTemplates({ childDir: partials, dir, extension })
    : {};
  const viewTemplates = compileTemplates({ childDir: views, dir, extension });

  if (Object.keys(viewTemplates).length === 0) {
    throw new Error(`No templates with extension ${extension} found in ${join(dir, views)}`);
  }

  const render: HandlebarsRender = (view, data = {}, options = {}) =>
    response({
      body: viewTemplates[view](data, { helpers, partials: partialTemplates }),
      status: 200 as const,
      ...options,
      headers: { ...defaultHeaders, ...options.headers },
    });

  return (next) => (req) => next({ ...req, render });
};
