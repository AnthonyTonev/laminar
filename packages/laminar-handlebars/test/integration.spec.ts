import { get, post, router, httpServer, start, stop } from '@ovotech/laminar';
import axios, { AxiosResponse } from 'axios';
import { join } from 'path';
import { handlebarsMiddleware } from '../src';

const axiosSnapshot = {
  config: expect.anything(),
  request: expect.anything(),
  headers: { date: expect.anything() },
};

const capitalizeName = (name: string): string => name.toUpperCase();

describe('Integration', () => {
  it('Should process response', async () => {
    const handlebars = handlebarsMiddleware({
      dir: join(__dirname, 'root'),
      helpers: { capitalizeName },
    });

    const server = httpServer({
      port: 8062,
      app: handlebars(
        router(
          get('/', ({ hbs }) => hbs('index')),
          post('/result', ({ hbs, body: { name } }) => hbs('result', { name }, { status: 201 })),
        ),
      ),
    });

    const api = axios.create({ baseURL: 'http://localhost:8062' });

    try {
      await start(server);

      expect(await api.get('/')).toMatchSnapshot<AxiosResponse>(axiosSnapshot);
      expect(await api.post('/result', { name: 'John Smith' })).toMatchSnapshot<AxiosResponse>(axiosSnapshot);
    } finally {
      await stop(server);
    }
  });
});
