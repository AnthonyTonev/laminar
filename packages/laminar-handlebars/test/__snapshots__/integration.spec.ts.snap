// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration Should process changing handlebars tempaltes with expiry: after modification 1`] = `
Object {
  "config": Anything,
  "data": "<html><body>Layout 2 <span>Generated 2</span></body></html>",
  "headers": Object {
    "connection": "close",
    "content-length": "59",
    "content-type": "text/html",
    "date": Anything,
  },
  "request": Anything,
  "status": 200,
  "statusText": "OK",
}
`;

exports[`Integration Should process changing handlebars tempaltes with expiry: inital 1`] = `
Object {
  "config": Anything,
  "data": "<html><body>Layout <span>Generated</span></body></html>",
  "headers": Object {
    "connection": "close",
    "content-length": "55",
    "content-type": "text/html",
    "date": Anything,
  },
  "request": Anything,
  "status": 200,
  "statusText": "OK",
}
`;

exports[`Integration Should process changing handlebars tempaltes with none: after modification 1`] = `
Object {
  "config": Anything,
  "data": "<html><body>Layout 2 <span>Generated 2</span></body></html>",
  "headers": Object {
    "connection": "close",
    "content-length": "59",
    "content-type": "text/html",
    "date": Anything,
  },
  "request": Anything,
  "status": 200,
  "statusText": "OK",
}
`;

exports[`Integration Should process changing handlebars tempaltes with none: inital 1`] = `
Object {
  "config": Anything,
  "data": "<html><body>Layout <span>Generated</span></body></html>",
  "headers": Object {
    "connection": "close",
    "content-length": "55",
    "content-type": "text/html",
    "date": Anything,
  },
  "request": Anything,
  "status": 200,
  "statusText": "OK",
}
`;

exports[`Integration Should process changing handlebars tempaltes with preload: after modification 1`] = `
Object {
  "config": Anything,
  "data": "<html><body>Layout <span>Generated</span></body></html>",
  "headers": Object {
    "connection": "close",
    "content-length": "55",
    "content-type": "text/html",
    "date": Anything,
  },
  "request": Anything,
  "status": 200,
  "statusText": "OK",
}
`;

exports[`Integration Should process changing handlebars tempaltes with preload: inital 1`] = `
Object {
  "config": Anything,
  "data": "<html><body>Layout <span>Generated</span></body></html>",
  "headers": Object {
    "connection": "close",
    "content-length": "55",
    "content-type": "text/html",
    "date": Anything,
  },
  "request": Anything,
  "status": 200,
  "statusText": "OK",
}
`;

exports[`Integration Should process response 1`] = `
Object {
  "config": Anything,
  "data": "<html>
  <body>
    <span>Index</span>
  </body>
</html>
",
  "headers": Object {
    "connection": "close",
    "content-length": "57",
    "content-type": "text/html",
    "date": Anything,
  },
  "request": Anything,
  "status": 200,
  "statusText": "OK",
}
`;

exports[`Integration Should process response 2`] = `
Object {
  "config": Anything,
  "data": "<html>
  <body>
    <span>name: JOHN SMITH</span>
    <span><span>has test John Smith</span>
    </span>
  </body>
</html>
",
  "headers": Object {
    "connection": "close",
    "content-length": "123",
    "content-type": "text/html",
    "date": Anything,
  },
  "request": Anything,
  "status": 201,
  "statusText": "Created",
}
`;
